### ISSUE TITLE
Borrowed classes get immediately freed after the first

### ISSUE THREAD

**Issue by daviditen** [chapel-lang/chapel#20516]
### Summary of Problem
When a borrowed class is created with

```chapel
var Log: borrowed Logger?;
Log = (new owned DefaultLogger()).borrow();
Log!.Println("I want to print from my DefaultLogger");
```

It is allocated and lives until the end of the scope. However if you allocate another class in the same way it is immediately `free`d and is not usable.

```chapel
Log = (new owned DefaultLogger2()).borrow();
Log!.Println("I want to print from my DefaultLogger2"); // potential crash here
```

~~These tests are hitting this problem, but often run anyway unless using a tool like valgrind:~~
~~[Error matching program output for classes/deitz/dispatch/test_dc1]~~
~~[Error matching program output for classes/deitz/inherit/test_inherit4b]~~
~~[Error matching program output for functions/varargs/varargOverrideGeneric]~~

This .future expects an error message but instead runs and accesses freed memory.
test/classes/delete-free/borrowed/reassignBorrowed.chpl

### Configuration Information

- Output of `chpl --version`:
chpl version 1.28.0 pre-release (c55c625893)
built with LLVM version 13.0.1
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: llvm *
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none *
CHPL_TASKS: fifo *
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: bundled *
CHPL_MEM: cstdlib *
CHPL_ATOMICS: cstdlib
CHPL_GMP: bundled *
CHPL_HWLOC: none
CHPL_RE2: none *
CHPL_LLVM: system *
CHPL_AUX_FILESYS: none

**Comment by mppf** [comment]
@daviditen - this is actually working as designed, but the part that is not working as designed is that the compiler should issue an error about it.

``` chapel
var Log: borrowed Logger?;
Log = (new owned DefaultLogger()).borrow();
Log!.Println("I want to print from my DefaultLogger");
```

The above is split-initializing `Log`, and as a result, is equivalent to `var Log: borrowed Logger? = (new owned DefaultLogger()).borrow();`. The rule for call-temporaries is that they live to the end of the block if they are in a variable initialization like this.

But, if you add another 

``` chapel
Log = (new owned DefaultLogger()).borrow();
```

Now the call temporary is not in a variable initialization. Instead of living to the end of the block, it is destroyed at the end of the statement.

**Comment by mppf** [comment]
(A moot question is why `new borrowed` did not have the same error as expanding it to the equivalent).

**Comment by daviditen** [comment]
Is a proper fix to these tests to add a separate `owned` variable for each `borrowed` instance? Or maybe just to make it use `owned` instead of `borrowed`?

**Comment by mppf** [comment]
> Is a proper fix to these tests to add a separate `owned` variable for each `borrowed` instance? Or maybe just to make it use `owned` instead of `borrowed`?

Yes I think having a separate variable for each (which, in effect, just stores the temporary), is called for, for these tests. Thanks.

**Comment by daviditen** [comment]
> Now the call temporary is not in a variable initialization. Instead of living to the end of the block, it is destroyed at the end of the statement.

That's very subtle. Is there a reason the initialization should be so different from another assignment in this way? I was thinking they both just had an owned temp that lived to the end of the block so would be the same.

**Comment by mppf** [comment]
It was a compromise solution.
 * For some I/O examples, it seemed important to deinit at end-of-statement.
 * For references to array slices, the RHS needs to be kept alive until the end of the block.

We did discuss it a fair amount in #11492 and #11534. However, IIRC, split-init did not exist at the time, and I agree that the combination makes it a bit more confusing.

**Comment by daviditen** [comment]
https://github.com/chapel-lang/chapel/pull/20529 fixed these tests to borrow from a declared owned instance and added a .future requesting an error message for this.
